### PR TITLE
8330388: Remove invokedynamic cache index encoding

### PR DESCRIPTION
**Pull request by matias9927**
Before [JDK-8307190](https://bugs.openjdk.org/browse/JDK-8307190), [JDK-8309673](https://bugs.openjdk.org/browse/JDK-8309673), and [JDK-8301995](https://bugs.openjdk.org/browse/JDK-8301995), invokedynamic operands needed to be rewritten to encoded values to better distinguish indy entries from other cp cache entries. The above changes now distinguish between entries with `to_cp_index()` using the bytecode, which is now propagated by the callers.

The encoding flips the bits of the index so the encoded index is always negative, leading to access errors if there is no matching decode call. These calls are removed with some methods adjusted to distinguish between indices with the bytecode. Verified with tier 1-5 tests. The changes show no issues when tested against libgraal.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8330388](https://bugs.openjdk.org/browse/JDK-8330388): Remove invokedynamic cache index encoding (**Bug** - P4)

### Reviewers
 * [Chris Plummer](https://openjdk.org/census#cjplummer) (@plummercj - **Reviewer**) ⚠️ Review applies to [87926aee](https://git.openjdk.org/jdk/pull/18819/files/87926aee8a297c2dbe917baf9f81c8c32568c886)
 * [Dean Long](https://openjdk.org/census#dlong) (@dean-long - **Reviewer**) ⚠️ Review applies to [87926aee](https://git.openjdk.org/jdk/pull/18819/files/87926aee8a297c2dbe917baf9f81c8c32568c886)
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18819/head:pull/18819` \
`$ git checkout pull/18819`

Update a local copy of the PR: \
`$ git checkout pull/18819` \
`$ git pull https://git.openjdk.org/jdk.git pull/18819/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18819`

View PR using the GUI difftool: \
`$ git pr show -t 18819`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18819.diff">https://git.openjdk.org/jdk/pull/18819.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18819#issuecomment-2061993552)

### CODE REVIEW

**Review by plummercj — APPROVED** [review]
SA changes look good.

File: src/hotspot/share/ci/ciEnv.cpp [hunk 1]
```diff
@@ -1499,21 +1497,20 @@ void ciEnv::record_call_site_method(Thread* thread, Method* adapter) {
...
-    oop appendix = cp->resolved_reference_from_indy(index);
+    oop appendix = cp->resolved_reference_from_indy(indy_index);
     {
       RecordLocation fp(this, "<appendix>");
       record_call_site_obj(thread, appendix);
     }
     // process the BSM
     int pool_index = indy_info->constant_pool_index();
-    BootstrapInfo bootstrap_specifier(cp, pool_index, index);
+    BootstrapInfo bootstrap_specifier(cp, pool_index, indy_index);
```
Comment (dean-long) [inline src/hotspot/share/ci/ciEnv.cpp:1513]: Why not just change the incoming parameter name to `index`?

File: src/hotspot/share/classfile/resolutionErrors.hpp [hunk 2]
```diff
@@ -59,7 +59,6 @@ class ResolutionErrorTable : AllStatic {
   // This function is used to encode an invokedynamic index to differentiate it from a
   // constant pool index.  It assumes it is being called with a index that is less than 0
```
Comment (dean-long) [inline src/hotspot/share/classfile/resolutionErrors.hpp:60]: Is this comment still correct?

File: src/hotspot/share/interpreter/bootstrapInfo.cpp [hunk 3]
```diff
@@ -74,7 +74,7 @@ bool BootstrapInfo::resolve_previously_linked_invokedynamic(CallInfo& result, TR
     Exceptions::wrap_dynamic_exception(/* is_indy */ true, CHECK_false);
     return true;
   } else if (indy_entry->resolution_failed()) {
-    int encoded_index = ResolutionErrorTable::encode_indy_index(ConstantPool::encode_invokedynamic_index(_indy_index));
+    int encoded_index = ResolutionErrorTable::encode_indy_index(_indy_index);
```
Comment (dean-long) [inline src/hotspot/share/interpreter/bootstrapInfo.cpp:77]: That's an improvement, from two levels of encoding to only one!

**Comment by dean-long** [comment]
Did you consider minimizing changes by leaving decode_invokedynamic_index/encode_invokedynamic_index calls in place, but having the implementations not change the value?

**Review by dean-long — APPROVED** [review]
@dougxc should check JVMCI changes.

File: src/jdk.internal.vm.ci/share/classes/jdk/vm/ci/hotspot/HotSpotConstantPool.java [hunk 4]
```diff
@@ -735,15 +717,7 @@ private static JavaType getJavaType(final Object type) {
...
-        int which; // interpretation depends on opcode
-        if (opcode == Bytecodes.INVOKEDYNAMIC) {
-            if (!isInvokedynamicIndex(rawIndex)) {
-                throw new IllegalArgumentException("expected a raw index for INVOKEDYNAMIC but got " + rawIndex);
-            }
-            which = rawIndex;
-        } else {
-            which = rawIndex;
-        }
+        int which = rawIndex;
```
Comment (gilles-duboscq) [inline src/jdk.internal.vm.ci/share/classes/jdk/vm/ci/hotspot/HotSpotConstantPool.java:720]: We could get rid of that intermediate variable now and just use `rawIndex` below.

**Comment by dougxc** [comment]
> @dougxc should check JVMCI changes.

Thanks for the heads up. I've asked @matias9927 to double check these changes against libgraal which should address any concerns about how this change impacts Graal.

**Review by coleenp — APPROVED** [review]
To borrow @shipilev's comment from a different PR, Good Riddance!

File: src/hotspot/share/classfile/resolutionErrors.hpp (hunk 2 above)
Comment (matias9927) [reply src/hotspot/share/classfile/resolutionErrors.hpp:60]: The last sentence is no longer valid, thanks for catching this!

**Comment by matias9927** [comment]
> Did you consider minimizing changes by leaving decode_invokedynamic_index/encode_invokedynamic_index calls in place, but having the implementations not change the value?

The intention from the start was to remove the encode/decode methods because they have been made unnecessary thanks to the changes mentioned in the description. As the author of the previously mentioned changes that overhauled the cpcache, this change should have been included in one of those PRs, but I must have forgotten! Leaving the calls in even if they did nothing would just make the code confusing and might become a red herring if other issues in the area come up.

File: src/hotspot/share/ci/ciEnv.cpp (hunk 1 above)
Comment (matias9927) [reply src/hotspot/share/ci/ciEnv.cpp:1513]: `indy_index` is used frequently even in functions that are only used in the context of invokedynamic. I think it helps with clarity.

**Review by coleenp — APPROVED** [review]
Still good!

**Comment by matias9927** [comment]
Thanks for the reviews @dean-long @gilles-duboscq @coleenp and @plummercj!
/integrate

**Pull request closed without merging**